### PR TITLE
importccl: support decompressing input files

### PR DESCRIPTION
**Pull request by dt**
This adds support for bzip and gzip decompression to IMPORT.

CSV or other text-based input formats are likely to benefit greatly from
compression -- repeated FKs or similar adjacent timestamps represented
in verbose textual forms for example likely mean many rows feature
repeated strings.

The decompression codec to use can be specified using the `decompress` option
to IMPORT. The values 'gzip' or 'bzip' are supported to explicitly specify a codec,
 as is the value 'auto' which attempts to guess the codec based on the filename,
matching the common extensions '.gz' or '.bz2' and `.bz`. This is the default
behavior if no option is explicitly specified, while 'none' can be used to
disable any attempts at decompression.

Release note (sql change): Support de-compressing IMPORT
files with gzip or bzip.

### CODE REVIEW

**Comment by bobvawter** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/ccl/importccl/read_import_proc.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26796#-LFIwBfDANv-mq41kJ25:-LFIwBfDANv-mq41kJ26:bfxu2f2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b4fa975871f5da6b604fdb90f4150e0f3b1ae1/pkg/ccl/importccl/read_import_proc.go#L165)):*
> ```Go
> }
> 
> func guessCompressionFromName(
> ```

How feasible is it to guess the type from the magic bytes that these formats start with?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26796)*

File: pkg/roachpb/io-formats.proto [hunk 1]
```diff
@@ -32,8 +32,17 @@ message IOFileFormat {
   optional CSVOptions csv = 2 [(gogoproto.nullable) = false];
   optional MySQLOutfileOptions mysql_out = 3 [(gogoproto.nullable) = false];
   optional PgCopyOptions pg_copy = 4 [(gogoproto.nullable) = false];
+
+  enum Compression {
+    Auto = 0;
+    None = 1;
+    Gzip = 2;
+    Bzip = 3;
```
Comment (mjibson) [inline pkg/roachpb/io-formats.proto]: Zip is in the standard library. Can we add that too? Or is it annoying because it could contain multiple files?

**Review by madelynnblue — APPROVED** [review]
LGTM. Ignore the zip comment if it's annoying to do because of the multi file thing.

**Comment by mjibson** [comment]
As a potential followup, here's the docs on `gsutil cp -z`:

```
  -z <ext,...>   Applies gzip content-encoding to any file upload whose
                 extension matches the -z extension list. This is useful when
                 uploading files with compressible content (such as .js, .css,
                 or .html files) because it saves network bandwidth and space
                 in Google Cloud Storage, which in turn reduces storage costs.

                 When you specify the -z option, the data from your files is
                 compressed before it is uploaded, but your actual files are
                 left uncompressed on the local disk. The uploaded objects
                 retain the Content-Type and name of the original files but are
                 given a Content-Encoding header with the value "gzip" to
                 indicate that the object data stored are compressed on the
                 Google Cloud Storage servers.

                 For example, the following command:

                   gsutil cp -z html -a public-read \
                     cattypes.html tabby.jpeg gs://mycats

                 will do all of the following:

                 - Upload the files cattypes.html and tabby.jpeg to the bucket
                   gs://mycats (cp command)
                 - Set the Content-Type of cattypes.html to text/html and
                   tabby.jpeg to image/jpeg (based on file extensions)
                 - Compress the data in the file cattypes.html (-z option)
                 - Set the Content-Encoding for cattypes.html to gzip
                   (-z option)
                 - Set the ACL for both files to public-read (-a option)
                 - If a user tries to view cattypes.html in a browser, the
                   browser will know to uncompress the data based on the
                   Content-Encoding header and to render it as HTML based on
                   the Content-Type header.

                 Note that if you download an object with Content-Encoding:gzip
                 gsutil will decompress the content before writing the local
                 file.
```

If you are interested and have time it would make sense to either support that or verify that http client reader correctly uncompresses them.

**Comment by dt** [comment]
Decided we're doing file names for now, not byte sniffing, and punting on zip file support until we decide what we want to do with multifile.

**Comment by bdarnell** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/ccl/importccl/import_stmt_test.go, line 488 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26796#-LFU6s4E-c_z_QfrFnU0:-LFU6s4E-c_z_QfrFnU1:bkrpphl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b4fa975871f5da6b604fdb90f4150e0f3b1ae1/pkg/ccl/importccl/import_stmt_test.go#L488)):*
> ```Go
> 	for i := range files {
> 		bzip[i] = bzipFile(t, dir, files[i])
> 		// if we don't have `bzip` on PATH, just skip those subtests.
> ```

Our builder image does not contain `bzip2` (unless I'm doing something wrong), so I think these tests will just get skipped on CI. We should probably add `bzip2` to the builder and have these tests fail instead of skip if it's not there.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26796)*

**Pull request merged**